corehq/motech/repeaters/tasks.py

```diff
@@ -305,19 +306,61 @@ def process_repeaters():
     Processes repeaters, instead of processing repeat records
     independently the way that ``check_repeaters()`` does.
     """
+    process_repeaters_lock = get_redis_lock(
+        PROCESS_REPEATERS_KEY,
+        timeout=None,  # Iterating repeaters forever is fine
+        name=PROCESS_REPEATERS_KEY,
+    )
+    if not process_repeaters_lock.acquire(blocking=False):
+        return
+
     metrics_counter('commcare.repeaters.process_repeaters.start')
-    for domain, repeater_id in iter_ready_repeater_ids():
+    try:
+        for repeater_id, lock_token in iter_repeater_id_tokens():
+            process_repeater(repeater_id, lock_token)
+    finally:
+        process_repeaters_lock.release()
+        metrics_counter('commcare.repeaters.process_repeaters.complete')
+
+
+def iter_repeater_id_tokens():
+    """
+    Iterate repeater_id-lock_token pairs until all ready repeat records
+    are processed.
+    """
+    while True:
+        metrics_counter('commcare.repeaters.process_repeaters.iter_once')
+        acquired_list = []
+        for repeater_id in iter_filtered_repeater_ids():
+            lock = RepeaterLock(repeater_id)
+            if acquired := lock.acquire():
+                yield repeater_id, lock.token
+            else:
+                metrics_counter('commcare.repeaters.process_repeaters.repeater_locked')
+            acquired_list.append(acquired)
+        if not acquired_list:
+            # No repeaters are ready, enabled, or not rate-limited
+            return
+        if not any(acquired_list):
+            # All repeaters are still processing. Sleep to allow at
+            # least one repeater to finish.
```

Comment:
nit: This wording implies that a repeater will finish during the wait, but that doesn't seem to be guaranteed. What do you think of
```suggestion
            # All repeaters are still processing. Sleep and hope that at
            # least one repeater finishes.
```

Reply:
I like it very much.